### PR TITLE
fix: Don't double-count metric events

### PR DESCRIPTION
**Pull request by simster7**
The K8s informer [may send multiple delete events](https://github.com/kubernetes/kubernetes/issues/22150) for the same deletion, causing the metrics service to double-count object deletions and resulting in negative workflow count numbers. To fix, keep a list of observed Workflow keys and ignore double events.

### CODE REVIEW

File: workflow/metrics/metrics.go [hunk 1]
```diff
@@ -47,14 +48,15 @@ type Metrics struct {
...
 
-func New(metricsConfig, telemetryConfig ServerConfig) Metrics {
-	metrics := Metrics{
+func New(metricsConfig, telemetryConfig ServerConfig) *Metrics {
+	metrics := &Metrics{
 		metricsConfig:      metricsConfig,
 		telemetryConfig:    telemetryConfig,
 		workflowsProcessed: newCounter("workflows_processed_count", "Number of workflow updates processed", nil),
 		workflowsByPhase:   getWorkflowPhaseGauges(),
+		workflows:          make(map[string]bool),
```
Comment (alexec) [inline workflow/metrics/metrics.go:59]: could you use the informer store?

Comment (simster7) [reply workflow/metrics/metrics.go:59]: Great suggestion. I'll take a look

Comment (simster7) [reply workflow/metrics/metrics.go:59]: Answer is no: the object is removed from the informer store before calling the handler function, making it impossible to distinguish if it's the first or second time the delete handler has been called.

File: workflow/metrics/metrics.go [hunk 2]
```diff
@@ -86,33 +88,44 @@ func (m Metrics) allMetrics() []prometheus.Metric {
...
+	m.WorkflowAdded(key, toPhase)
 }
 
-func (m Metrics) WorkflowDeleted(phase v1alpha1.NodePhase) {
+func (m *Metrics) WorkflowDeleted(key string, phase v1alpha1.NodePhase) {
+	if !m.workflows[key] {
+		return
+	}
 	if _, ok := m.workflowsByPhase[phase]; ok {
+		delete(m.workflows, key)
```
Comment (alexec) [inline workflow/metrics/metrics.go:114]: move one line higher?

File: workflow/metrics/metrics.go [hunk 3]
```diff
@@ -86,33 +88,44 @@ func (m Metrics) allMetrics() []prometheus.Metric {
 	return allMetrics
 }
 
-func (m Metrics) WorkflowAdded(phase v1alpha1.NodePhase) {
+func (m *Metrics) WorkflowAdded(key string, phase v1alpha1.NodePhase) {
+	if m.workflows[key] {
+		return
+	}
 	if _, ok := m.workflowsByPhase[phase]; ok {
+		m.workflows[key] = true
```
Comment (alexec) [inline workflow/metrics/metrics.go:96]: move one line higher?

File: workflow/metrics/metrics.go (hunk 2 above)
Comment (simster7) [reply workflow/metrics/metrics.go]: Yes sorry, this was lingering from when I used to have the check on the `if` statement

File: workflow/metrics/metrics.go (hunk 3 above)
Comment (simster7) [reply workflow/metrics/metrics.go]: Yes sorry, this was lingering from when I used to have the check on the `if` statement

File: workflow/metrics/metrics.go (hunk 1 above)
Comment (alexec) [reply workflow/metrics/metrics.go:59]: ok, I think we should have two changes, essentially this:

```
type workflowTracker struct {
  lock sync.Mutex
  workflows map[string]boo
}
```

That object can be responsible for tracking workflows - but before you do that - can we ask @jessesuen if he has a suggestion?

**Pull request merged**